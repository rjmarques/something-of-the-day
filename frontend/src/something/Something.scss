@import "../theme.scss";

.something-element {
  display: flex;
  flex-direction: column;
  height: 100vh;
  min-width: 360px;
  max-width: 780px;
  margin: 0px auto;
  padding: 40px 15px 0px 15px;

  .main-content {
    flex: 1 0 auto;

    > * {
      margin-bottom: 40px;
    }

    header {
      display: flex;
      flex-flow: column;
      font-family: BirdsofParadise;
      font-size: 45px;

      @include for-size(laptop-small-up) {
        font-size: 65px;
      }

      .title {
        align-self: center;
        margin-bottom: 30px;
      }

      img {
        align-self: flex-end;
        border: 5px solid white;
        border-radius: 5px;
        box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2),
          0 6px 20px 0 rgba(0, 0, 0, 0.19);
        height: 120px;
        opacity: 0.8;
        width: 120px;

        @include for-size(tablet-up) {
          height: 148px;          
          width: 148px;
        }
      }
    }

    .something-box {
      background-color: #383e4a;
      box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2),
        0 6px 20px 0 rgba(0, 0, 0, 0.19);
      padding: 50px;
      border-radius: 5px;
      min-height: 180px;

      p {
        font-size: 20px;
        font-family: Helvetica;
        white-space: pre-wrap;
      }
    }

    .button-area {
      .something-button {
        width: 250px;
        font-size: 25px;
      }
    }
  }

  footer {
    color: #68707f;
    flex-shrink: 0;
    font-family: monospace;
    margin-bottom: 20px;

    .icons {
      margin-bottom: 10px;
    }

    span {
      font-size: 12px;
    }

    a {
      color: #b0796e;
      text-decoration: none;

      &:hover {
        opacity: 0.8;
      }
    }

    img {
      filter: invert(56%) sepia(4%) saturate(3210%) hue-rotate(322deg)
        brightness(91%) contrast(97%);
      height: 24px;
      margin: 0px 20px;
      width: 24px;
    }
  }
}
