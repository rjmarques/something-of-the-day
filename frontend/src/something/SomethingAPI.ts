import { BASE_URL, Get } from "../shared/services/http";

export interface ISomething {
  Id: number;
  CreatedAt: string;
  Text: string;
}

export async function GetSomething(): Promise<ISomething> {
  const url = `${BASE_URL}/something`;

  const something: ISomething = await Get(url);
  return something;
}
