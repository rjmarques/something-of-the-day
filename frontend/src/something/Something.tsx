import React, { Component } from "react";
import { Button, Spinner } from "reactstrap";

import "./Something.scss";
import "../assets/img/github.svg";

import { ISomething } from "./SomethingAPI";

export interface IProps {
  getSomething: () => Promise<ISomething>;
}

export interface IState {
  loading: boolean;
  something?: ISomething;
}

export class Something extends Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);

    this.state = {
      loading: false
    };
  }

  public componentDidMount() {
    this.updateSomething();
  }

  public render() {
    return (
      <div className="something-element">
        <div className="main-content">
          <header>
            <div className="title">Something of the day </div>
            <div>
              <img
                src={require("../assets/img/ascihappy.png")}
                alt="happy"
              ></img>
            </div>
          </header>
          <div className="something-box">
            <p>{this.state.something?.Text}</p>
            {this.state.loading ? (
              <Spinner type="grow" color="info" />
            ) : (
              <noscript />
            )}
          </div>
          <div className="button-area">
            <Button
              className="something-button"
              onClick={this.updateSomething}
              color="primary"
            >
              Ooh La La{" "}
              <span role="img" aria-label="kiss">
                ðŸ’‹
              </span>
            </Button>
          </div>
        </div>
        <footer>
          <div className="icons">
            <span>
              <a href="https://github.com/rjmarques" target="blank">
                <img src={require("../assets/img/github.svg")} alt="github" />
              </a>
            </span>
            <span>
              <a href="https://twitter.com/lgst_something" target="blank">
                <img src={require("../assets/img/twitter.svg")} alt="twitter" />
              </a>
            </span>
          </div>
          <div>
            <span>
              Â© Copyright {new Date().getFullYear()}{" "}
              <a href="https://ricardomarques.dev" target="blank">
                Ricardo Marques
              </a>
            </span>
          </div>
        </footer>
      </div>
    );
  }

  public updateSomething = async () => {
    try {
      if (this.state.loading) {
        return;
      }

      this.setState({ loading: true });

      const something = await this.props.getSomething();
      if (something) {
        this.setState({ something });
      }
    } catch (error) {
      console.log(error);
    } finally {
      this.setState({ loading: false });
    }
  };
}

export default Something;
