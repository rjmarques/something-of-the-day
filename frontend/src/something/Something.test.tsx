import React from "react";
import renderer from "react-test-renderer";
import Something from "./Something";
import { ISomething } from "./SomethingAPI";

it("renders correctly", async () => {
  const mockSomething: ISomething = {
    Id: 1,
    CreatedAt: "today",
    Text: "Let's put a smile on that face!"
  };
  const getSomething = jest
    .fn()
    .mockReturnValue(Promise.resolve(mockSomething));

  const component = renderer.create(<Something getSomething={getSomething} />);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  await Promise.resolve();

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
