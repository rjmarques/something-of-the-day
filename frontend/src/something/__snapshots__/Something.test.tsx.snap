// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="something-element"
>
  <div
    className="main-content"
  >
    <header>
      <div
        className="title"
      >
        Something of the day 
      </div>
      <div>
        <img
          alt="happy"
          src="ascihappy.png"
        />
      </div>
    </header>
    <div
      className="something-box"
    >
      <p />
      <div
        className="spinner-grow text-info"
        role="status"
      >
        <span
          className="sr-only"
        >
          Loading...
        </span>
      </div>
    </div>
    <div
      className="button-area"
    >
      <button
        aria-label={null}
        className="something-button btn btn-primary"
        onClick={[Function]}
        type="button"
      >
        Ooh La La
         
        <span
          aria-label="kiss"
          role="img"
        >
          ðŸ’‹
        </span>
      </button>
    </div>
  </div>
  <footer>
    <div
      className="icons"
    >
      <span>
        <a
          href="https://github.com/rjmarques"
          target="blank"
        >
          <img
            alt="github"
            src={
              Object {
                "ReactComponent": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                },
                "__esModule": true,
                "default": "github.svg",
              }
            }
          />
        </a>
      </span>
      <span>
        <a
          href="https://twitter.com/lgst_something"
          target="blank"
        >
          <img
            alt="twitter"
            src={
              Object {
                "ReactComponent": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                },
                "__esModule": true,
                "default": "twitter.svg",
              }
            }
          />
        </a>
      </span>
    </div>
    <div>
      <span>
        Â© Copyright 
        2020
         
        <a
          href="https://ricardomarques.dev"
          target="blank"
        >
          Ricardo Marques
        </a>
      </span>
    </div>
  </footer>
</div>
`;

exports[`renders correctly 2`] = `
<div
  className="something-element"
>
  <div
    className="main-content"
  >
    <header>
      <div
        className="title"
      >
        Something of the day 
      </div>
      <div>
        <img
          alt="happy"
          src="ascihappy.png"
        />
      </div>
    </header>
    <div
      className="something-box"
    >
      <p>
        Let's put a smile on that face!
      </p>
      <noscript />
    </div>
    <div
      className="button-area"
    >
      <button
        aria-label={null}
        className="something-button btn btn-primary"
        onClick={[Function]}
        type="button"
      >
        Ooh La La
         
        <span
          aria-label="kiss"
          role="img"
        >
          ðŸ’‹
        </span>
      </button>
    </div>
  </div>
  <footer>
    <div
      className="icons"
    >
      <span>
        <a
          href="https://github.com/rjmarques"
          target="blank"
        >
          <img
            alt="github"
            src={
              Object {
                "ReactComponent": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                },
                "__esModule": true,
                "default": "github.svg",
              }
            }
          />
        </a>
      </span>
      <span>
        <a
          href="https://twitter.com/lgst_something"
          target="blank"
        >
          <img
            alt="twitter"
            src={
              Object {
                "ReactComponent": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                },
                "__esModule": true,
                "default": "twitter.svg",
              }
            }
          />
        </a>
      </span>
    </div>
    <div>
      <span>
        Â© Copyright 
        2020
         
        <a
          href="https://ricardomarques.dev"
          target="blank"
        >
          Ricardo Marques
        </a>
      </span>
    </div>
  </footer>
</div>
`;
